tmpl/blog/introducing-solo5.md

```diff
@@ -0,0 +1,125 @@
+
+* **Memory layout and protection.** Hardware "features" like
+  page-level write protection must be exposed by the lowest layer for
+  techniques like memory tracing to be performed.  Also,
+  software-level strategies like address space layout randomization
+  require cooperation of this lowest layer.
+
+* **Low-level device interfacing.** As individual devices (e.g., NICs)
+  gain virtualization capabilities, the lowest-software layer is an
+  obvious place to interface directly with hardware.
+
+* **Threads/events.** The low-level code must ensure that device I/O
+  is asynchronous and/or fits with the higher-level synchronization
+  primitives.
+
+The most popular existing code providing this low-level kernel layer
+for is called Mini-OS.  Mini-OS was (I believe) originally written as
+a vehicle to demonstrate the paravirtualized interface offered by Xen
+for people to have a reference to port their kernels to and as a base
+for new kernel builders to build specialized Xen domains.  Mini-OS is
+a popular base for MirageOS, ClickOS, and other unikernels.  Other
+software that implements a unikernel base include rumprun and OSv.
+
+I built Solo5 from scratch (rather than adapting Mini-OS, for example)
+primarily as an educational (and fun!) exercise to explore and really
+understand the role of the low-level kernel layer in a unikernel.  To
+provide applications, Solo5 supports the Mirage stack.  It is my hope
+that Solo5 can be a useful base for others; even if only at this point
+to run some Mirage applications on KVM/QEMU!
+
+### Solo5: Building a Unikernel Base From Scratch
+
+At a high level, there are roughly 3 parts to building a unikernel
+base that runs on KVM/QEMU and supports Mirage:
+
+* **Typical kernel hardware initialization.** The kernel must know how
+  to loading things into memory at the desired locations, and prepare
+  the processor to operate in the correct mode (e.g., 64-bit).  Unlike
+  typical kernels, most setup is one-time and simplified.  The kernel
+  must set up a memory map, stack, interrupt vectors, and provide
+  primitives for basic memory allocation.  At its simplest, a
+  unikernel base kernel does not need to worry about user address
+  spaces, threads, or many other things typical kernels need.
+
+* **Interact with `virtio` devices.** `virtio` is a paravirtualized
+  device standard supported by some hypervisors, including KVM/QEMU
+  and Virtualbox.  As far as devices go, `virtio` devices are simple:
+  I was able to write (very simple/unoptimized) `virtio` drivers for
+  Solo5 drivers from scratch in C.  At some point it may be
+  interesting to write them in OCaml like the Xen device drivers in
+  Mirage, but for someone who doesn't know OCaml (like me) a simple C
+  implementation seemed like a good first step.  I should note that
+  even though the drivers themselves are written in C, to connect with
+  Mirage Solo5 does include some OCaml code to call out to the
+  drivers.
```

Comment:
`to connect with Mirage Solo5 does include some OCaml code to call out to the drivers.` -> `Solo5 does include some OCaml code to call out to the drivers so it can connect with Mirage.`